File: app/src/main/java/com/zaze/demo/component/notification/service/KeepLiveService.java
package com.zaze.demo.component.notification.service;

import android.app.Notification;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.os.IBinder;

import androidx.core.app.NotificationCompat;

import com.zaze.demo.R;
import com.zaze.demo.component.application.AppActivity;

/**
 * Description :
 *
 * @author : ZAZE
 * @version : 2016-11-03 - 17:25
 */
public class KeepLiveService extends Service {

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this);
        PendingIntent contentIntent = PendingIntent.getActivity(this, 0,
                new Intent(this, AppActivity.class), PendingIntent.FLAG_UPDATE_CURRENT);
        builder.setContentIntent(contentIntent);
        //设置小图标
        builder.setSmallIcon(R.mipmap.ic_launcher);
        //设置通知标题
        builder.setContentTitle("通知标题");
        //设置通知内容
        builder.setContentText("通知内容");
        builder.setTicker("Test Ticker");
        Notification notification = builder.build();
        startForeground(0, notification);
        stopForeground(true);
        // --------------------------------------------------
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }
}

File: app/src/main/java/com/zaze/demo/debug/KeepLiveService.kt
package com.zaze.demo.debug

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log
import com.zaze.utils.log.ZLog
import com.zaze.utils.log.ZTag
import java.io.File
import java.io.FileOutputStream
import java.nio.channels.FileLock

/**
 * Description :
 *
 * @author : ZAZE
 * @version : 2020-05-12 - 13:25
 */
class KeepLiveService : Service() {
    private var thread = LockThread()
    private var lockFile: File? = null
    private var keepApp: Pair<String, String>? = null


    companion object {
        const val PACKAGE_NAME = "packageName"
        const val LOCK_FILE = "lockFile"
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    override fun onCreate() {
        super.onCreate()
        Log.i(this.packageName, "KeepLiveService : onCreate")
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val startCompatibility = super.onStartCommand(intent, flags, startId)
        if (intent == null) {
            return startCompatibility
        }
        val appPackage = intent.getStringExtra(PACKAGE_NAME)
        val lockFile = intent.getStringExtra(LOCK_FILE)
        if (appPackage.isNullOrEmpty() || lockFile.isNullOrEmpty()) {
            return startCompatibility
        }
        keepApp = Pair(appPackage, lockFile)
        Log.i(this.packageName, "KeepLiveService : onStartCommand >> $thread")
        Log.i(this.packageName, "KeepLiveService : ${thread.state}")
        when (thread.state) {
            Thread.State.RUNNABLE -> {
                return startCompatibility
            }
            Thread.State.NEW -> {
            }
            Thread.State.TERMINATED -> {
                thread = LockThread()
            }
            else -> {
                return startCompatibility
            }
        }
        if (thread.isInterrupted) {
            thread = LockThread()
        }
        thread.start()
        return startCompatibility
    }

    override fun onDestroy() {
        Log.i(this.packageName, "KeepLiveService : onDestroy")
        super.onDestroy()
        thread.interrupt()
    }

    inner class LockThread : Thread() {
        override fun run() {
            super.run()
            keepApp?.let {
                Log.i("plock", "plock ${it.first} >> ${it.second} start")
//                lock(File(it.second))
                Log.i("plock", "plock ${it.first} >> ${it.second}  end")
            }
        }

        private fun lock(lockFile: File): FileLock? {
            try {
                val fos = FileOutputStream(lockFile)
                val fl = fos.channel.lock()
                if (fl.isValid) {
                    ZLog.i(ZTag.TAG, "LockThread lock " + lockFile.absolutePath + " SUC!")
                    return fl
                }
            } catch (e: Exception) {
                ZLog.w(ZTag.TAG, "LockThread lock " + lockFile.absolutePath + " FAIL!", e)
            }
            return null
        }
    }
}